### PR TITLE
Move `SnippetViewSet` menu registration attributes and methods to base `ViewSet` class

### PR DESCRIPTION
**Pull request by laymonage**
Part of #10740.

This introduces the `WagtailMenuRegisterable` and `WagtailMenuRegisterableGroup` classes, which encapsulates menu item registration and can be used outside of `ViewSet`s.

If we don't want this to be in the base `ViewSet` class, we can also create another class e.g. `WagtailAdminViewSet`.

_Please check the following:_

-   [x] Do the tests still pass?[^1]
-   [x] Does the code comply with the style guide?
    -   [x] Run `make lint` from the Wagtail root.
-   [x] For Python changes: Have you added tests to cover the new/fixed behaviour?
    - Added tests for the base `ViewSet` class, will write more for the `ModelViewSet` in upcoming PRs when we make further changes to `ModelViewSet`.

[^1]: [Development Testing](https://docs.wagtail.org/en/latest/contributing/developing.html#testing)
[^2]: [Browser and device support](https://docs.wagtail.org/en/latest/contributing/developing.html#browser-and-device-support)
[^3]: [Accessibility Target](https://docs.wagtail.org/en/latest/contributing/developing.html#accessibility-targets)

### CODE REVIEW

File: wagtail/admin/viewsets/__init__.py [hunk 1]
```diff
@@ -10,13 +11,17 @@ def __init__(self):
     def populate(self):
         for fn in hooks.get_hooks("register_admin_viewset"):
             viewset = fn()
-            if isinstance(viewset, (list, tuple)):
-                for vs in viewset:
-                    self.register(vs)
-            else:
-                self.register(viewset)
```
Comment (laymonage) [inline wagtail/admin/viewsets/__init__.py:17]: This was added in 9d1110d349a43205407df0b5e73f33bd0a3a031e, but it's no longer used since the viewset registration was moved to `register_snippet` in 6e1e0240ef3ada234f6ee9d7db2584c141d29f78.

File: wagtail/admin/viewsets/base.py [hunk 2]
```diff
@@ -15,6 +18,9 @@ class ViewSet:
     #: A name for this viewset, used as the default URL prefix and namespace.
     name = None
 
+    #: The icon to use across the views.
+    icon = ""
```
Comment (laymonage) [inline wagtail/admin/viewsets/base.py:22]: Not sure if this should be here because the base `ViewSet` doesn't have any preconfigured views, but I thought it'd be nice to standardise this here and so that the `menu_icon` uses this by default.

I extracted this into a separate commit just now, sorry for making further changes after sending the review request!

File: wagtail/admin/viewsets/model.py [hunk 3]
```diff
@@ -102,6 +115,25 @@ def delete_view(self):
             header_icon=self.icon,
         )
 
+    @cached_property
+    def menu_label(self):
+        return self.model_opts.verbose_name_plural.title()
```
Comment (laymonage) [inline wagtail/admin/viewsets/model.py:120]: Will add tests for this and the default `name` in a separate PR when we focus on `ModelViewSet`. For now, the changes in this PR are mostly focused on the base `ViewSet` class, so I didn't add any `ModelViewSet`-specific tests.

**Review by laymonage — COMMENTED** [review]
I think this is now ready for review.

**Review by gasman — APPROVED** [review]
All looks good to me! I particularly like the idea of having `WagtailMenuRegisterable` mixed in to the base class but disabled by default, so that you only need `add_to_admin_menu = True` to get a menu item for any viewset - and also picking up the app label as the default menu label for a ModelViewSetGroup. (I expect it'll be pretty unusual for a developer not to specify the label explicitly, but a group of related models from the same app is exactly the case where a friendly default feels like the natural thing to do...) Details like those make me confident that we've got a really nice API design going on here!
